xcrypto/hash.go

```diff
@@ -175,12 +203,44 @@ func (h *evpHash) BlockSize() int {
 }
 
 func (h *evpHash) Sum(b []byte) []byte {
+	h.initialize()
 	digest := make([]byte, h.size)
 	C.memcpy(h.ctx2, h.ctx, C.size_t(h.ctxSize))
 	h.final(h.ctx2, digest)
 	return append(b, digest...)
 }
 
+// Clone returns a new evpHash object that is a deep clone of itself.
+// The duplicate object contains all state and data contained in the
+// original object at the point of duplication.
+func (h *evpHash) Clone() hash.Hash {
+	h.initialize()
+	cloned := &evpHash{
+		init:      h.init,
+		update:    h.update,
+		final:     h.final,
+		blockSize: h.blockSize,
+		size:      h.size,
+		ctxSize:   h.ctxSize,
+	}
+	cloned.ctx = C.malloc(C.size_t(h.ctxSize))
+	cloned.ctx2 = C.malloc(C.size_t(h.ctxSize))
+	if cloned.ctx == nil || cloned.ctx2 == nil {
```

Comment:
No need to check if `malloc` failed, the cgo runtime will panic if that happens,